Separating particle velocity and spatial initialization

gsgall opened the pull request:
In order to improve code reuse for particle initialization particle velocity and spatial location have been separated.

In my own usage I have run into cases where I want to use a specific spatial initialization but vary the velocity initialization. This enables that capability and provides the system with a much greater degree of flexibility.

EDIT: Ref. #87

gsgall committed:
Removing velocity initialization from the particle initializer class

- This is the first step in the change that will allow the user to
  specify a velocity space initialization indpendent from the physical
  space initialization.

- Addressing issue #87

gsgall committed:
Removed particle velocity sampling capabailities addressing #87

- A new velocity initializer has been removed from the particle
  initializer so that someone can utilize the same spatial particle
  initialization with a different velocity initialization

gsgall committed:
Updating input files for the new particle intialization

gsgall committed:
bug fix for the new object

gsgall committed:
Fixing bug in the initial velocity initializer implementation

gsgall committed:
Updating testing files to accomodate the new velocity initializer scheme

- Input files were updated to account for the new initialization method

- The constant distribution file was removed since it is no longer
  needed and was not a proper use of the distribution system

- Updated the study used for testing to account for these changes as
  well

gsgall committed:
Adding documentation for the new velocity initialization capabilities

gsgall committed:
Cleaning up some documentation and comments within class objects and
applying formatting

gsgall committed:
Removing white spaces from input files

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3118737/), step Sync to remote on 09ae766 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/salamander/docs/PRs/97/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3118742/), step Generate coverage on 09ae766 wanted to post the following:


## Coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>762d38</th>
      <th colspan=4>#97 09ae76</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>93.90%</td>
      <td>93.78%</td>
      <td>-0.12%</td>
      <td>94.87%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>646</td>
      <td>663</td>
      <td>+17</td>
      <td>37</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>42</td>
      <td>44</td>
      <td>+2</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/salamander/docs/PRs/97/coverage_diff)

[Full coverage report](https://mooseframework.inl.gov/salamander/docs/PRs/97/coverage)



This comment will be updated on new commits.

gsgall committed:
Updating the documentation in the lieberman example

- This example case no longer uses the Distribution system in order to
  initialize particle velocities. Since I cannot build the docs without
  complining TMAP and Cardinal as well this broke.

simopier commented on doc/content/source/VelocityInitializerBase.md:
```diff
@@ -0,0 +1,3 @@
+# VelocityInitializerBase
+
+`VelocityInitializerBase` is the abstract class from which all velocity initializers should be derived. The method `getParticleVelocities` should be called by a [ParticleInitializerBase.md]  and will return a list of `Points` which are the initial particle velocities.
```
```suggestion
`VelocityInitializerBase` is the abstract class from which all velocity initializers should be derived. The method `getParticleVelocities` should be called by a [ParticleInitializerBase.md] and will return a list of `Points` which are the initial particle velocities.
```

simopier commented on doc/content/source/VelocityInitializerBase.md:
Should this really be at the source level? 
Seeing where the `VelocityInitializerBase` class sits, I think it should be under userobjects.

gsgall replied:
Yes I believe that it should be in the user objects. I will go ahead and move it.

simopier commented on src/userobjects/ConstantVelocityInitializer.C:
```diff
@@ -0,0 +1,46 @@
+{
+  auto params = VelocityInitializerBase::validParams();
+  params.addClassDescription(
+      "Provides particles with velocities sampled from a user provided list of velocities");
```
```suggestion
      "Provides particles with velocities sampled from a user provided list of velocities.");
```

simopier commented on doc/content/source/userobjects/ConstantVelocityInitializer.md:
```diff
@@ -0,0 +1,18 @@
+## Overview
+
+This velocity initializer provides particles with velocities from the provided list.
+When more velocities are requested from this initializer than are provided as input, the list of velocities will be repeated until the total number of velocities requested have been provided.
```
Should a warning be returned when that's the case?

gsgall replied:
I don't think so. This is how I would like this feature to work. This allows us to use the same code to give all of the particles in the system the same velocity or specify a velocity for each particle that is getting created. This object is also mostly going to be used for supporting testing so I would be fine moving this to a testing object as well if you have concerns about this being confusing for the user.

simopier commented on doc/content/verification_validation_examples/plasma/lieberman.md:
```diff
@@ -129,9 +129,13 @@ For this simulation, only the x component of the electric field is being used. I
+The `UserObjects` block is where particles are created and the rules for how the particle velocity is updated by the fields are defined. The `stepper` is a [LeapFrogStepper.md] which updates a particles' velocity based on the value of the electric fields at the location at which the particle exists.
+The `initializer` defines the rules for how particles are placed in the mesh and the `velocity_initializer` defines how their velocities are initialized.
+The [UniformGridParticleInitializer.md] places particles evenly throughout the mesh.
+In this case, 100 particles are placed on the mesh with uniform spacing between them, and they are weighted so that this particle distribution will approximate the specified argon ion number density, `n`.
```
So you removed 
```
The last parameter, `velocity_distributions`, tells the system which distributions that should be sampled from when initializing particle velocity data.
```
?

gsgall replied:
Yes the velocity initializer class will handle all of the sampling. In the future we will probably need to add an initializer class that enables sampling from distributions. This will probably be added when I get to adding the initial DSMC code since the test case does use an initialization list this. But for now none of the code nor tests require this ability.

simopier commented on include/userobjects/VelocityInitializerBase.h:
```diff
@@ -0,0 +1,46 @@
+   */
+  virtual const std::vector<Point> getParticleVelocities(const size_t num_samples) const = 0;
+
+  /**
+   * Unused methods
+   */
+  ///@{
+  virtual void initialize() override final {}
+  virtual void finalize() override final {}
+  virtual void execute() override final {}
+  ///@}
```
I know you have the same in particleInitBase, and I know I asked before, but can you remind me why those are needed here?

gsgall replied:
These are pure virtual functions in the GeneralUserObject so we must override them in order for this to be a concrete class.

simopier commented on src/userobjects/ConstantVelocityInitializer.C:
```diff
@@ -0,0 +1,46 @@
+
+  std::generate(velocities.begin(),
+                velocities.end(),
+                [this, i = 0]() mutable { return _velocities[i++ % _velocities.size()]; });
```
Can you help me understand what that does? especially the `[this, i = 0]() mutable`.

loganharbour replied:
This is equivalent to

```cpp
for (const auto i : index_range(_velocities))
  velocities[i] = _velocities[i % _velocities.size()];
```

simopier commented on test/tests/benchmarking/lieberman.i:
```diff
@@ -93,26 +93,24 @@ num_samples = ${fparse num_elem + 1}
+    velocities = '0 0 0'
+  []
+
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/raybcs/reflection_base.i:
```diff
@@ -7,10 +7,15 @@
+    type = ConstantVelocityInitializer
+  []
+
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/userobjects/particle_initializer/bounding_box/elements/initializer_base.i:
```diff
@@ -21,18 +21,16 @@ charge_density = 2
+    velocities = '0 0 0'
+  []
+
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/userobjects/particle_initializer/bounding_box/errors/errors.i:
```diff
@@ -20,22 +20,20 @@ charge_density = 2
+    type = ConstantVelocityInitializer
+    velocities = '0 0 0'
+  []
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/userobjects/particle_initializer/per_element/elements/initializer_base.i:
```diff
@@ -49,20 +49,17 @@ charge_density = 0
+    type = ConstantVelocityInitializer
+    velocities = '0 0 0'
+  []
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/userobjects/particle_initializer/per_element/errors/errors.i:
```diff
@@ -13,20 +13,18 @@
-[]
 
-[UserObjects]
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/userobjects/particle_initializer/per_element/errors/tests:
```diff
@@ -27,14 +27,6 @@
       expect_err = 'Range check failed for parameter UserObjects/initializer/mass'
       detail = 'tries to give particles a negative mass'
     []
-    [too_few_distributions]
-      type = 'RunException'
-      input = 'errors.i'
-      allow_test_objects = true
-      cli_args = "UserObjects/initializer/velocity_distributions='zero zero'"
-      expect_err = 'You must provide 3 distributions, one for each velocity component.'
-      detail = 'does not provide enough distributions to sample for velocity initialization.'
-    []
```
Why not update this to provide too few velocities and ensure it does indeed create enough (and maybe creates a warning to ensure the user is aware of that).

gsgall replied:
This case would already be handled by the framework since the ConstantVelocityInitializer takes a vector of libMesh points as input the InputParameters object will give you an error if you do not provide something that is a valid collection of points.

simopier commented on test/tests/userobjects/particle_initializer/uniform_grid/uniform_grid.i:
```diff
@@ -30,7 +27,7 @@
     charge = 1
```
I can't provide a suggestion to replace `[initializer]` with `[particle_initializer]` (too far from your changes), but please make sure you do so. I think it helps distinguish with `[velocity_initializer]`

simopier commented on test/tests/userobjects/particle_quantity_accumulation/number_density_accumulator.i:
```diff
@@ -32,18 +32,18 @@
+    type = ConstantVelocityInitializer
+    velocities = '0 0 0'
+  []
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/userobjects/particle_quantity_accumulation/simple_potential_solve.i:
```diff
@@ -45,18 +45,19 @@
+    velocities = '0 0 0'
+  []
+
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/userobjects/particle_stepper/boris_stepper/boris_parallel_acceleration.i:
```diff
@@ -92,7 +96,7 @@
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/userobjects/particle_stepper/boris_stepper/boris_projectile_motion.i:
```diff
@@ -90,10 +90,15 @@
+    velocities = '10 10 0'
+  []
+
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/userobjects/particle_stepper/boris_stepper/cyclotron_motion.i:
```diff
@@ -91,10 +91,15 @@
+    velocities = '1 0 0'
+  []
+
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/userobjects/particle_stepper/boris_stepper/e_cross_b.i:
```diff
@@ -91,10 +91,14 @@
+    type = ConstantVelocityInitializer
+    velocities = '1 0 0'
+  []
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/userobjects/particle_stepper/leapfrog_stepper/parallel_acceleration.i:
```diff
@@ -55,10 +55,14 @@
+    type = ConstantVelocityInitializer
+    velocities = '0 0 0'
+  []
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/userobjects/particle_stepper/leapfrog_stepper/projectile_motion.i:
```diff
@@ -56,10 +56,15 @@
+    velocities = '10 10 0'
+  []
+
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/userobjects/particle_stepper/simple_stepping/simple_stepping.i:
```diff
@@ -14,15 +14,19 @@
+                  0.5 0.9 10.8'
+  []
+
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/utils/auxaccumulator/charge_accumulator_base.i:
```diff
@@ -13,9 +13,14 @@
+    velocities = '0 0 0'
+  []
+
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier commented on test/tests/utils/auxaccumulator/simple_potential_solve_aux.i:
```diff
@@ -39,6 +39,11 @@
+
   [stepper]
     type = TestSimpleStepper
   []
```
`particle_` here too - I can't add it myself.

simopier commented on test/tests/utils/variable_sampler/variable_sampler.i:
```diff
@@ -34,10 +34,15 @@
+    velocities = '1 1 1'
+  []
+
   [initializer]
```
```suggestion
  [particle_initializer]
```

simopier reviewed (changes requested):
I think this is a good improvement. I only have some minor suggestions.

loganharbour reviewed (commented):
No concerns from me on the code side

gsgall committed:
Apply suggestions from code review

Co-authored-by: Pierre-Clement Simon <pierreclement.simon@gmail.com>

gsgall committed:
Moved the VelocityInitializerBase object to the proper location

gsgall committed:
Updated input files and updated parameters to use 'particle_initializer'

- The TestInitializedPICStudy now requires the name particle_initializer
  instead of initializer for consitency.

- Updated all test input files to account for this change as well.

gsgall committed:
Explicitly setting the type of the loop index in lambda function

gsgall commented:
Your requested changes should be addressed now @simopier

gsgall committed:
Fixing detail in Lieberman example documentation

gsgall committed:
Updating parameter name in documentation

simopier commented on doc/content/verification_validation_examples/plasma/lieberman.md:
```diff
@@ -1,4 +1,4 @@
-# Lieberman Ion Wall Losses
+#n Lieberman Ion Wall Losses
```
```suggestion
# Lieberman Ion Wall Losses
```

cticenhour replied:
I concur with this change.

simopier reviewed (changes requested):
Thanks for this, @gsgall. There's just one small typo that somehow came up, and the rest looks good!

cticenhour commented on include/userobjects/ConstantVelocityInitializer.h:
```diff
@@ -0,0 +1,32 @@
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
+// confinement for Numerical Design, Engineering & Research,
```
```suggestion
//* confinement for Numerical Design, Engineering & Research,
```

cticenhour commented on include/userobjects/ParticleInitializerBase.h:
```diff
@@ -1,4 +1,5 @@
-//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic confinement for Numerical Design, Engineering & Research,
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
+// confinement for Numerical Design, Engineering & Research,
```
```suggestion
//* confinement for Numerical Design, Engineering & Research,
```

cticenhour commented on include/userobjects/VelocityInitializerBase.h:
```diff
@@ -0,0 +1,46 @@
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
+// confinement for Numerical Design, Engineering & Research,
```
```suggestion
//* confinement for Numerical Design, Engineering & Research,
```

cticenhour commented on src/userobjects/BoundingBoxParticleInitializer.C:
```diff
@@ -1,4 +1,5 @@
-//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic confinement for Numerical Design, Engineering & Research,
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
+// confinement for Numerical Design, Engineering & Research,
```
```suggestion
//* confinement for Numerical Design, Engineering & Research,
```

cticenhour commented on src/userobjects/ConstantVelocityInitializer.C:
```diff
@@ -0,0 +1,46 @@
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
+// confinement for Numerical Design, Engineering & Research,
```
```suggestion
//* confinement for Numerical Design, Engineering & Research,
```

cticenhour commented on src/userobjects/ParticleInitializerBase.C:
```diff
@@ -1,4 +1,5 @@
-//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic confinement for Numerical Design, Engineering & Research,
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
+// confinement for Numerical Design, Engineering & Research,
```
```suggestion
//* confinement for Numerical Design, Engineering & Research,
```

cticenhour commented on src/userobjects/PerElementParticleInitializer.C:
```diff
@@ -1,4 +1,5 @@
-//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic confinement for Numerical Design, Engineering & Research,
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
+// confinement for Numerical Design, Engineering & Research,
```
```suggestion
//* confinement for Numerical Design, Engineering & Research,
```

cticenhour commented on src/userobjects/UniformGridParticleInitializer.C:
```diff
@@ -1,4 +1,5 @@
-//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic confinement for Numerical Design, Engineering & Research,
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
+// confinement for Numerical Design, Engineering & Research,
```
```suggestion
//* confinement for Numerical Design, Engineering & Research,
```

cticenhour commented on src/userobjects/VelocityInitializerBase.C:
```diff
@@ -0,0 +1,35 @@
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
+// confinement for Numerical Design, Engineering & Research,
```
```suggestion
//* confinement for Numerical Design, Engineering & Research,
```

cticenhour commented on src/userobjects/VelocityInitializerBase.C:
```diff
@@ -0,0 +1,35 @@
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
+// confinement for Numerical Design, Engineering & Research,
+//* A multiphysics application for modeling plasma facing components *
+// https://github.com/idaholab/salamander
```
```suggestion
//* https://github.com/idaholab/salamander
```

cticenhour commented on test/include/userobjects/TestPlacedParticleInitializer.h:
```diff
@@ -1,4 +1,5 @@
-//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic confinement for Numerical Design, Engineering & Research,
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
+// confinement for Numerical Design, Engineering & Research,
```
```suggestion
//* confinement for Numerical Design, Engineering & Research,
```

cticenhour commented on test/src/userobjects/TestInitializedPICStudy.C:
```diff
@@ -1,4 +1,5 @@
-//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic confinement for Numerical Design, Engineering & Research,
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
+// confinement for Numerical Design, Engineering & Research,
```
```suggestion
//* confinement for Numerical Design, Engineering & Research,
```

cticenhour commented on test/src/userobjects/TestPlacedParticleInitializer.C:
```diff
@@ -1,4 +1,5 @@
-//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic confinement for Numerical Design, Engineering & Research,
+//* This file is part of SALAMANDER: Software for Advanced Large-scale Analysis of MAgnetic
+// confinement for Numerical Design, Engineering & Research,
```
```suggestion
//* confinement for Numerical Design, Engineering & Research,
```

cticenhour reviewed (changes requested):
Lots of minor suggestions. Otherwise looks good.

gsgall committed:
Apply suggestions from code review

Co-authored-by: Pierre-Clement Simon <pierreclement.simon@gmail.com>
Co-authored-by: Casey Icenhour <cticenhour@gmail.com>

simopier reviewed (approved):
Thank you @gsgall for this contribution and @cticenhour and @loganharbour for the reviews!
